### PR TITLE
[EVM] Add bridging interface to EVM contract

### PR DESCRIPTION
**Pull request by sisyphusSmiling**
Related: #5663

Adds pre-1.0 bridging interfaces to EVM contract as [suggested](https://github.com/onflow/flow-go/pull/5664#issuecomment-2057950122) by @turbolent in #5664 

It looks like a number of tests are still failing, though I can't get to the bottom of why. In particular, `fvm_test.TestEVM` cases appear to be failing on failed contract declaration even though this PR includes updates to alias NFT & FT contracts.

### CODE REVIEW

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [inline fvm/evm/stdlib/contract.cdc:318]: Not sure if we should allow withdraw for non-coa accounts? 
I can see we force the reference to be COA but why not just exposing it on the COA resource

File: fvm/fvm_test.go
Comment (ramtinms) [inline fvm/fvm_test.go:3065]: are the changes in this file expected?

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [inline fvm/evm/stdlib/contract.cdc:299]: nit: we may want to unify the feeprovider interface for all methods, 
Currently for some operations like withdraw tokens we don't collect any extra fee and we just charge the payer of the transaction.

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [inline fvm/evm/stdlib/contract.cdc:599]: do we have to use a borrow approach here, instead of keeping the bridge accessor as a field in the EVM and directly using it?

File: fvm/evm/stdlib/contract.cdc
Comment (sisyphusSmiling) [reply fvm/evm/stdlib/contract.cdc:318]: This issue here is that, the COA ultimately needs to execute a `safeTransfer` call to some recipient against some EVM contract - here an ERC721 - to escrow the EVM asset. Within the scope of the EVM contract, how do we know:
- the escrow recipient address?
- the EVM contract address to target the call against?
If we want to know these values in scope, we'd need to further engrain the bridge into the EVM contract and I feel that given the critical nature of the EVM contract it would be best to avoid that.

To decouple the circular dependency between the EVM contract and bridge logic contracts, I thought it best to  pass the COA reference on to an implementation of interfaces defined within this contract and handle the resolution of those values within a contract that more in scope of bridge configuration. The implementation this reference is getting passed to is defined [here](https://github.com/onflow/flow-evm-bridge/blob/main/cadence/contracts/bridge/FlowEVMBridgeAccessor.cdc#L46).

File: fvm/evm/stdlib/contract.cdc
Comment (sisyphusSmiling) [reply fvm/evm/stdlib/contract.cdc:299]: I'm not sure I follow. In this case, the `feeProvider` is providing a fee that is calculated by the bridge contracts. This could be a reference on a Vault, or, as currently implemented in the initial transactions, a `ScopedFTProvider` which enables the caller to set a withdrawal limit (see [this call](https://github.com/onflow/flow-evm-bridge/blob/main/cadence/transactions/bridge/nft/bridge_nft_to_evm.cdc#L80L83))

File: fvm/evm/stdlib/contract.cdc
Comment (sisyphusSmiling) [reply fvm/evm/stdlib/contract.cdc:599]: Hmm, what do you mean? Adding a field to the EVM contract wouldn't be an upgradable change, so not sure of the specifics required for that update. Or are you suggesting something else?

**Comment by turbolent** [comment]
Nice work! BTW, I only suggested to have this on master and not only on the Cadence 1.0 feature branch, so the EVM code in the branches is the same, besides Cadence 1.0 compatibility

File: fvm/fvm_test.go
Comment (sisyphusSmiling) [reply fvm/fvm_test.go:3065]: I'm not sure if they're expected, but I didn't find a way around them given the inclusion of `NonFungibleToken` and `FungibleToken` in the EVM contract. What are the outsized implications of changing this file?

Comment (ramtinms) [reply fvm/fvm_test.go:3065]: I see we derive the addresses from the chain ID. looks good.

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [reply fvm/evm/stdlib/contract.cdc:299]: I meant for flow tokens we don't collect any fees for bridging but we do charge for other fungible tokens, not 100% related to this PR, we can iterate over it later.

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [reply fvm/evm/stdlib/contract.cdc:318]: I see, so basically you pass the reference to the COA, so the bridge can decide what calls to do.

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [inline fvm/evm/stdlib/contract.cdc:327]: I have difficulty understanding why we pass self.address() for to here. but we are going to talk about it

Comment (sisyphusSmiling) [reply fvm/evm/stdlib/contract.cdc:327]: The `to` value here tells the bridge the EVMAddress to which transfer the bridged asset. The idea is that depositing an NFT/FT to a COA would deposit the EVM representation of the asset to the caller on the EVM side. While we could enable a caller to define that themselves and basically execute a bridge + transfer, feedback pointed to this being a/ the more common case and b/ simpler & easier to grok.

For those that do wish to define a distinct address, they can call the bridge contract directly - see [here](https://github.com/onflow/flow-evm-bridge/blob/main/cadence/contracts/bridge/FlowEVMBridge.cdc#L189L194)

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [reply fvm/evm/stdlib/contract.cdc:599]: Gio is smart, this is the best way to do it at the moment.

File: fvm/evm/stdlib/contract.cdc
Comment (ramtinms) [reply fvm/evm/stdlib/contract.cdc:327]: right the self here refer to the COA.

**Review by ramtinms — APPROVED** [review]
Nice work 👏

**Comment by sisyphusSmiling** [comment]
Thank you @ramtinms! FYI I added an event and setter in [78efce2](https://github.com/onflow/flow-go/pull/5677/commits/78efce223123bca30ae3b1d8ccaa37f64898dbba) based on feedback from you and @janezpodhostnik. I don't believe we can emit events in interfaces pre-1.0, but I plan on emitting `BridgeAccessorUpdated` within the new `BridgeRouter.setBridgeAccessor` method pre-condition in the 1.0 version.

**Review by janezpodhostnik — APPROVED** [review]

**Comment by turbolent** [comment]
CI fails because test is already broken on master currently

**Comment by sisyphusSmiling** [comment]
> CI fails because test is already broken on master currently

Is this an action item for this PR or just a statement saying this is ready for merge?

**Comment by turbolent** [comment]
@sisyphusSmiling 
> > CI fails because test is already broken on master currently
> 
> Is this an action item for this PR or just a statement saying this is ready for merge?

A test on master is currently broken, so all PRs against master fail too.
I opened #5737 to try and fix it.
See https://discord.com/channels/613813861610684416/1230603448802742362

Once master is fixed, master should be merged into this PR, and then it can be merged once all tests pass.

**Comment by sisyphusSmiling** [comment]
@turbolent thanks for the guidance, master has been merged and it looks like CI is now passing.

**Pull request merged**
